### ISSUE TITLE
Tectonic fails to work with process substitution

### ISSUE THREAD

**Issue by kozross** [tectonic-typesetting/tectonic#186]
I have a file ``lecture.tex`` made with Beamer, and I wanted to generate a 'presentation version' and a 'handout version'. To do this, I use ``sed``, in the following manner:

```
sed 's/\\documentclass{beamer}/\\documentclass[handout]{beamer}/' lecture.tex
```

To avoid having to generate an intermediate file, I decided to use process substitution:

```
tectonic <(sed 's/\\documentclass{beamer}/\\documentclass[handout]{beamer}/' lecture.tex)
```

However, Tectonic doesn't like that very much, complaining that

```
warning: input seek failed
caused by: Illegal seek (os error 29)
error: something bad happened inside TeX; its output follows:

===============================================================================
(63
LaTeX2e <2016/03/31>
Babel <3.9r> and hyphenation patterns for 83 language(s) loaded.
)
! Emergency stop
*** (job aborted, no legal \end found)
No pages of output.
Transcript written on 63.log.
===============================================================================

error: the TeX engine had an unrecoverable error
caused by: *** (job aborted, no legal \end found)
```

I have attached the relevant ``lecture.tex`` file to aid replication efforts, [as a zip file](https://github.com/tectonic-typesetting/tectonic/files/2172251/lecture.zip) because Github hates LaTeX or something.

**Comment by pkgw** [comment]
Thanks for the report!

My guess is that the problem is as follows. When process substitution is used in this way, the actual command line that is run is something like `tectonic /proc/fd/44`, where the shell has set up file descriptor number 44 to be a Unix pipe between the two programs. As far as Tectonic is concerned, this means that it is running with an input file like any other, and it attempts to reread the input multiple times in order to loop the engine. This behavior isn't supported by Unix pipes, however, hence the error.

This suggests a workaround. If Tectonic is using stdin as its input, it will buffer the data in memory to avoid this problem. So I expect the following should work:

```
sed 's/.../.../' lecture.tex |tectonic
```
I think that on Unix-y system you can test whether an open file descriptor points at something like a pipe, so it might be possible for Tectonic to realize that it is reading an un-seekable input file even when invoked with something like shell process substitution. In which case, it could automatically decide to use the same buffering logic that it uses for standard input. Perhaps there's even already a Rust crate for doing that kind of check in a cross-platform manner?

Finally, if this issue comes up in a variety of circumstances, it might make sense to add a command-line option indicating that the main input path should be buffered in memory without doing any OS-specific kinds of tests.

**Comment by kozross** [comment]
Thank you for that very detailed explanation! Is there a reason why Tectonic can't buffer the file in any case? The memory load for that can't be very large, and it would speed up larger file processing by not having to read it multiple times from disk.

**Comment by kozross** [comment]
Also, the command you provided doesn't work. When you try to do that, I get an error like this:

```
error: The following required arguments were not provided:
    <INPUT>

USAGE:
    tectonic <INPUT> --chatter <LEVEL> --format <PATH> --outfmt <FORMAT> --pass <PASS>

For more information try --help
```

**Comment by pkgw** [comment]
Sorry, you need to run it as `tectonic -` to get it to read from stdin with the buffering. As I recall, I thought (and continue to think) that it would be confusing for new users if an argument-less invocation just caused the program to seem to hang there. (More experienced users might guess it was waiting to read TeX commands, but that's not super obvious.)

I would be reluctant to have Tectonic buffer the input by default: such inputs *shouldn't* be large, but they might be non-trivially big in some use cases, and there seems to be some interest in Tectonic playing a role of a light-weight, low-resource TeX engine that such a default would cut against, I think. If there's a reliable way to detect that we're reading from unseekable streams, like pipes, I'd prefer to go that route at the moment ... but I'm certainly open to argument.

**Comment by kozross** [comment]
I'm no expert on stream-seeking, so I can't really comment on that part, but I think your reasoning is sound. It may be good to have the documentation mention something like:

```
If INPUT is -, then Tectonic will read from stdin.
```

I know it's standard UNIXy behaviour, but not everything does it, and documenting this feature would be a great help (since I certainly didn't know that it could do that).

**Comment by pkgw** [comment]
Yes, I thought the same thing when checking why the no-argument command didn't do what I initially expected. Can I persuade you to submit a PR adding the text? (Only change I'd suggest would be "standard input" rather than "stdin".) I can do so if not, but I like it when people can get some credit in the history for their improvements.

**Comment by kozross** [comment]
What would I need to modify to make these changes? I don't know Rust at all.

**Comment by pkgw** [comment]
You'd need to modify [this piece of code](https://github.com/tectonic-typesetting/tectonic/blob/master/src/bin/tectonic.rs#L206) to add a [long_help](https://docs.rs/clap/2.32.0/clap/struct.Arg.html#method.long_help) call. But to compile Tectonic out you'll have to install the Rust toolchain and the needed development packages, etc., so this isn't worth bothering unless that's something you feel like taking on! If you've got better things to do with your time I can definitely do it instead.

**Comment by kozross** [comment]
I think I'd prefer not to - I have a lot of other work, and I'm not knowledgeable about Rust at all, so it'd take me time to read up on everything.

**Comment by pkgw** [comment]
Fixed in #189.
